core/crypto/src/signature.rs

```diff
@@ -556,6 +558,62 @@ impl<'de> serde::Deserialize<'de> for SecretKey {
 #[derive(Clone)]
 pub struct Secp256K1Signature([u8; 65]);
 
+impl Secp256K1Signature {
+    pub fn check_signature_values(&self, reject_upper: bool) -> bool {
+        let mut r_bytes = [0u8; 32];
+        r_bytes.copy_from_slice(&self.0[0..32]);
+        let r = U256::from(r_bytes);
+
+        let mut s_bytes = [0u8; 32];
+        s_bytes.copy_from_slice(&self.0[32..64]);
+        let s = U256::from(s_bytes);
+
+        let secp256k1_n = U256::from([
+            0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+            0xff, 0xfe, 0xba, 0xae, 0xdc, 0xe6, 0xaf, 0x48, 0xa0, 0x3b, 0xbf, 0xd2, 0x5e, 0x8c,
+            0xd0, 0x36, 0x41, 0x41,
+        ]);
+
+        if reject_upper {
+            let secp256k1_half_n = secp256k1_n / U256::from(2);
+            // Reject upper range of s values (ECDSA malleability)
+            if s > secp256k1_half_n {
+                return false;
+            }
```

Comment:
We can always check `s` against just one upper bound: `s < if reject_upper { secp256k1_half_n+1 } else { secp256k1_n }` (where `secp256k1_half_n+1` should be a static).

Reply:
Ahh ok yeah I get it.

Reply:
https://github.com/near/nearcore/pull/4380/commits/38fc3626d696ebacbb9f03c5bb35bfb9412b7feb